### ISSUE TITLE
Building a Better Profanity Detection Library with scikit-learn

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4050]
Building a Better Profanity Detection Library with scikit-learn<br>
<br>
http://bit.ly/2Sxm7Br<br>
<br>
About 0 Minutes<br>
<br>
<div><div>
<section name="2bec"><div><hr></div>
<div>
<div>
<h2 name="1e79" id="1e79">Why existing libraries are uninspiring and how I built a better one.</h2>
<div>
<div><a href="https://towardsdatascience.com/@victorczhou?source=post_header_lockup" dir="auto"><img src="https://cdn-images-1.medium.com/fit/c/100/100/1*nGx5t_wfmnmey8B7Ngmv5w.png" alt="Go to the profile of Victor Zhou"></a></div>
<div>
<div>
<a href="https://towardsdatascience.com/@victorczhou" dir="auto">Victor Zhou</a><span></span>
</div>
<div>
<time datetime="2019-02-04T15:05:29.825Z">Feb 4</time><span></span><span title="6 min read"></span><span><span><svg width="15" height="15"></svg></span></span>
</div>
</div>
</div>
<p name="6095" id="6095">A few months ago, I needed a way to detect profanity in user-submitted text strings:</p>
<figure name="a263" id="a263"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*i2fk4aGvplR7le_3PPajAA.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*i2fk4aGvplR7le_3PPajAA.png">
</div>
</div>
<figcaption>This shouldn’t be that hard, right?</figcaption></figure><p name="b7fe" id="b7fe">I ended up building and releasing my own library for this purpose called <a href="https://github.com/vzhou842/profanity-check" rel="noopener" target="_blank">profanity-check</a>:</p>
<div name="f76c" id="f76c">
<a href="https://github.com/vzhou842/profanity-check" title="https://github.com/vzhou842/profanity-check"><strong>vzhou842/profanity-check</strong><br><em>A fast, robust Python library to check for offensive language in strings. - vzhou842/profanity-check</em>github.com</a><a href="https://github.com/vzhou842/profanity-check"></a>
</div>
<p name="2dce" id="2dce">Of course, before I did that, I looked in the <a href="https://pypi.org/" rel="noopener" target="_blank">Python Package Index</a> (PyPI) for any existing libraries that could do this for me. The only half decent results for the search query “profanity” were:</p>
<ul>
<li name="c22f" id="c22f">
<a href="https://pypi.org/project/profanity/" rel="noopener" target="_blank">profanity</a> (the ideal package name)</li>
<li name="dedb" id="dedb">
<a href="https://pypi.org/project/better-profanity/" rel="noopener" target="_blank">better-profanity</a>: <em>“Inspired from package </em><a href="https://github.com/ben174/profanity" rel="nofollow noopener" target="_blank"><em>profanity</em></a><em> of </em><a href="https://github.com/ben174" rel="nofollow noopener" target="_blank"><em>Ben Friedland</em></a><em>, this library is much faster than the original one.”</em>
</li>
<li name="80ae" id="80ae">
<a href="https://pypi.org/project/profanityfilter/" rel="noopener" target="_blank">profanityfilter</a> (has 31 Github stars, which is 30 more than most of the other results have)</li>
<li name="1c23" id="1c23">
<a href="https://pypi.org/project/profanity-filter/" rel="noopener" target="_blank">profanity-filter</a> (uses Machine Learning, enough said?!)</li>
</ul>
<p name="8538" id="8538">Third-party libraries can sometimes be sketchy, though, so I did my due diligence on these 4 results.</p>
<h3 name="f96b" id="f96b">profanity, better-profanity, and profanityfilter</h3>
<p name="dea7" id="dea7">After a quick dig through the <code>profanity</code> repository, I found a file named <a href="https://github.com/ben174/profanity/blob/master/profanity/data/wordlist.txt" rel="noopener" target="_blank">wordlist.txt</a>:</p>
<figure name="3226" id="3226"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*0lTbmHR5WE7HZ8wCvLpqtg.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*0lTbmHR5WE7HZ8wCvLpqtg.png">
</div>
</div>
<figcaption>Sorry this image of profanities is so big…</figcaption></figure><p name="de59" id="de59">The entire <code>profanity</code> library is just a wrapper over this list of 32 words! <code>profanity</code> detects profanity simply by looking for one of these words.</p>
<p name="d7e6" id="d7e6">To my dismay, <code>better-profanity</code> and <code>profanityfilter</code> both took the same approach:</p>
<ul>
<li name="868f" id="868f">
<code>better-profanity</code> uses <a href="https://github.com/snguyenthanh/better_profanity/blob/master/better_profanity/profanity_wordlist.txt" rel="noopener" target="_blank">a 140-word wordlist</a>
</li>
<li name="4c80" id="4c80">
<code>profanityfilter</code> uses <a href="https://github.com/areebbeigh/profanityfilter/blob/master/profanityfilter/data/badwords.txt" rel="noopener" target="_blank">a 418-word wordlist</a>
</li>
</ul>
<p name="3b48" id="3b48">This is bad because <strong>profanity detection libraries based on wordlists are extremely subjective. </strong>For example, <code>better-profanity</code>'s wordlist includes the word “suck.” Are you willing to say that any sentence containing the word “suck” is profane? Furthermore, any hard-coded list of bad words will inevitably be incomplete — do you think <code>profanity</code>'s 32 bad words are the only ones out there?</p>
</div>
<div><figure name="7fad" id="7fad"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*n5OWj4WEPkGexXO28_yteg.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*n5OWj4WEPkGexXO28_yteg.png">
</div>
</div>
<figcaption>Fucking Blue Shells. source: <a href="https://xkcd.com/290/" rel="noopener" target="_blank">xkcd</a></figcaption></figure></div>
<div>
<p name="0e60" id="0e60">Having already ruled out 3 libraries, I put my hopes on the 4th and final one: <code>profanity-filter</code>.</p>
<h3 name="24af" id="24af">profanity-filter</h3>
<p name="5b37" id="5b37"><code>profanity-filter</code> uses Machine Learning! Sweet!</p>
<p name="9903" id="9903">Turns out, it’s <strong><em>really</em> </strong>slow. Here’s a benchmark I ran in December 2018 comparing (1) <code>profanity-filter</code>, (2) my library <code>profanity-check</code>, and (3) <code>profanity</code> (the one with the list of 32 words):</p>
<figure name="4c5e" id="4c5e"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*KRJEl4YHfSTk9PmmScIcUA.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*KRJEl4YHfSTk9PmmScIcUA.png">
</div>
</div>
<figcaption>A human could probably do this faster than profanity-filter can</figcaption></figure><p name="f181" id="f181">I needed to be able to perform many predictions in real time, and <code>profanity-filter</code> was not even close to being fast enough. But hey, maybe this is a classic tradeoff of accuracy for speed, right?</p>
<p name="0309" id="0309">Nope.</p>
<figure name="180d" id="180d"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*LYOeGE6vTXTAKhJ_W1fZgQ.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*LYOeGE6vTXTAKhJ_W1fZgQ.png">
</div>
</div>
<figcaption>At least profanity-filter is not dead last this time</figcaption></figure><p name="4a66" id="4a66">None of the libraries I’d found on PyPI met my needs, so I built my own.</p>
<h3 name="29d7" id="29d7">Building profanity-check, Part 1: Data</h3>
<p name="b6b9" id="b6b9">I knew that I wanted <code>profanity-check</code> to base its classifications on data to avoid being subjective <em>(read: to be able to say I used Machine Learning)</em>. I put together a combined dataset from two publicly-available sources:</p>
<ul>
<li name="2189" id="2189">the “Twitter” dataset from <a href="https://github.com/t-davidson/hate-speech-and-offensive-language/tree/master/data" rel="noopener" target="_blank">t-davidson/hate-speech-and-offensive-language</a>, which contains tweets scraped from Twitter.</li>
<li name="d855" id="d855">the “Wikipedia” dataset from <a href="https://www.kaggle.com/c/jigsaw-toxic-comment-classification-challenge" rel="noopener" target="_blank">this Kaggle competition</a> published by Alphabet’s <a href="https://conversationai.github.io/" rel="noopener" target="_blank">Conversation AI</a> team, which contains comments from Wikipedia’s talk page edits.</li>
</ul>
<p name="9699" id="9699">Each of these datasets contains text samples hand-labeled by humans through crowdsourcing sites like <a href="https://www.figure-eight.com/" rel="noopener" target="_blank">Figure Eight</a>.</p>
<p name="2637" id="2637">Here’s what my dataset ended up looking like:</p>
<figure name="a475" id="a475"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*Bw_we8cbs-WOpWXOCxzSTg.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*Bw_we8cbs-WOpWXOCxzSTg.png">
</div>
</div>
<figcaption>Combined = Tweets + Wikipedia</figcaption></figure><blockquote name="0a51" id="0a51">The Twitter dataset has a column named <code>class</code> that’s 0 if the tweet contains hate speech, 1 if it contains offensive language, and 2 if it contains neither. I classified any tweet with a <code>class</code> of 2 as “Not Offensive” and all other tweets as “Offensive.”</blockquote>
<blockquote name="68d7" id="68d7">The Wikipedia dataset has several binary columns (e.g. <code>toxic</code> or <code>threat</code>) that represent whether or not that text contains that type of toxicity. I classified any text that contained <em>any </em>of the types of toxicity as “Offensive” and all other texts as “Not Offensive.”</blockquote>
<h3 name="1e9b" id="1e9b">Building profanity-check, Part 2: Training</h3>
<p name="a635" id="a635">Now armed with a cleaned, combined dataset (which you can <a href="https://github.com/vzhou842/profanity-check/blob/master/profanity_check/data/clean_data.csv" rel="noopener" target="_blank">download here</a>), I was ready to train the model!</p>
<blockquote name="c828" id="c828">I’m skipping over how I cleaned the dataset because, honestly, it’s pretty boring— if you’re interested in learning more about preprocessing text datasets check out <a href="https://machinelearningmastery.com/clean-text-machine-learning-python/" rel="noopener" target="_blank">this</a> or <a href="https://medium.com/@datamonsters/text-preprocessing-in-python-steps-tools-and-examples-bf025f872908" target="_blank">this</a>.</blockquote>
<figure name="248e" id="248e"><div>
<div></div>
<div></div>
</div>
<figcaption>Are you also surprised the code is so short? Apparently <a href="https://scikit-learn.org/" rel="noopener" target="_blank">scikit-learn</a> does everything.</figcaption></figure><p name="073a" id="073a">Two major steps are happening here: (1) vectorization and (2) training.</p>
<h4 name="6db4" id="6db4">Vectorization: Bag of Words</h4>
<p name="ac3e" id="ac3e">I used <code>scikit-learn</code>'s <a href="https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html" rel="noopener" target="_blank">CountVectorizer</a> class, which basically turns any text string into a vector by counting how many times each given word appears. This is known as a <a href="https://en.wikipedia.org/wiki/Bag-of-words_model" rel="noopener" target="_blank">Bag of Words</a> (BOW) representation. For example, if the only words in the English language were <code>the</code>, <code>cat</code>, <code>sat</code>, and <code>hat</code>, a possible vectorization of the sentence <code>the cat sat in the hat</code> might be:</p>
<figure name="936d" id="936d"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*sbnts1u_QFB_V-X5DSC3pg.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*sbnts1u_QFB_V-X5DSC3pg.png">
</div>
</div>
<figcaption>“the cat sat in the hat” -&gt; [2, 1, 1, 1, 1]</figcaption></figure><p name="7f81" id="7f81">The <code>???</code> represents any unknown word, which for this sentence is <code>in</code>. Any sentence can be represented in this way as counts of <code>the</code>, <code>cat</code>, <code>sat</code>, <code>hat</code>, and <code>???</code>!</p>
<figure name="482e" id="482e"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*-wONWZDab2gNQP3Rfdpt_A.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*-wONWZDab2gNQP3Rfdpt_A.png">
</div>
</div>
<figcaption>A handy reference table for the next time you need to vectorize “cat cat cat cat cat”</figcaption></figure><p name="bead" id="bead">Of course, there are far more words in the English language, so in the code above I use the <code>fit_transform()</code> method, which does 2 things:</p>
<ul>
<li name="36c3" id="36c3">
<strong>Fit: </strong>learns a vocabulary by looking at all words that appear in the dataset.</li>
<li name="f389" id="f389">
<strong>Transform</strong>: turns each text string in the dataset into its vector form.</li>
</ul>
<h4 name="2b51" id="2b51">Training: Linear SVM</h4>
<p name="f367" id="f367">The model I decided to use was a Linear Support Vector Machine (SVM), which is implemented by <code>scikit-learn</code>'s <a href="https://scikit-learn.org/stable/modules/generated/sklearn.svm.LinearSVC.html" rel="noopener" target="_blank">LinearSVC</a> class. <a href="https://medium.com/machine-learning-101/chapter-2-svm-support-vector-machine-theory-f0812effc72" target="_blank">This</a> and <a href="https://www.svm-tutorial.com/2014/11/svm-understanding-math-part-1/" rel="noopener" target="_blank">this</a> are good introductions if you don’t know what SVMs are.</p>
<blockquote name="a553" id="a553">The <a href="https://scikit-learn.org/stable/modules/generated/sklearn.calibration.CalibratedClassifierCV.html" rel="noopener" target="_blank">CalibratedClassifierCV</a> in the code above exists as a wrapper to give me the <code>predict_proba()</code> method, which returns a probability for each class instead of just a classification. You can pretty much just ignore it if that last sentence made no sense to you, though.</blockquote>
<p name="2a5d" id="2a5d">Here’s one (simplified) way you could think about why the Linear SVM works: during the training process, the model learns which words are “bad” and how “bad” they are because those words appear more often in offensive texts. <strong>It’s as if the training process is picking out the “bad” words for me</strong>, which is much better than using a wordlist I write myself!</p>
<p name="e15c" id="e15c">A Linear SVM combines the best aspects of the other profanity detection libraries I found: it’s fast enough to run in real-time yet robust enough to handle many different kinds of profanity.</p>
<h3 name="2521" id="2521">Caveats</h3>
<p name="2e11" id="2e11">That being said, <code>profanity-check</code> is far from perfect. Let me be clear: take predictions from <code>profanity-check</code> with a grain of salt because<strong> it makes mistakes. </strong>For example, its not good at picking up less common variants of profanities like “f4ck you” or “you b1tch” because they don’t appear often enough in the training data. You’ll never be able to detect <em>all </em>profanity (people will come up with new ways to evade filters), but <code>profanity-check</code> does a good job at finding most.</p>
<h3 name="2d00" id="2d00">profanity-check</h3>
<p name="8b4c" id="8b4c"><code>profanity-check</code> is open source and available on PyPI! To use it, simply</p>
<pre name="91fa" id="91fa">$ pip install profanity-check</pre>
<p name="d4ef" id="d4ef">How could <code>profanity-check</code> be even better? Feel free to reach out or comment with any thoughts or suggestions!</p>
<div name="3923" id="3923">
<a href="https://github.com/vzhou842/profanity-check" title="https://github.com/vzhou842/profanity-check"><strong>vzhou842/profanity-check</strong><br><em>A fast, robust Python library to check for offensive language in strings. - vzhou842/profanity-check</em>github.com</a><a href="https://github.com/vzhou842/profanity-check"></a>
</div>
</div>
</div></section><section name="5b29"><div><hr></div>
<div><div><p name="b915" id="b915"><em>Also posted on </em><a href="https://victorzhou.com/blog/better-profanity-detection-with-scikit-learn/" rel="noopener" target="_blank"><em>victorzhou.com</em></a><em>.</em></p></div></div></section>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via victorzhou.com https://victorzhou.com<br>
<br>
May 23, 2019 at 11:54AM
